lib/db/set.go

```diff
@@ -326,12 +325,12 @@ func (s *Snapshot) Availability(file string) []protocol.DeviceID {
 	return av
 }
 
-func (s *Snapshot) DebugGlobalVersions(file string) VersionList {
+func (s *Snapshot) DebugGlobalVersions(file string) *dbproto.VersionList {
```

Comment:
I think there should also be a non-proto `VersionList` type, as it's used in an exposed method here.

Just using this occurrence  (and the only I found) for the general question:  
What's the intent of making the generated types internal, resp. what's the intent for how/where they should be used - e.g. only internally, never in exposed methods?  
I was initially worried about the same issues raised around the recent proposal to make all/most of `./lib/...` internal: Those internal proto types being used in methods that are "public-ish", e.g. used by the iOS app. However then I quickly saw that we use the non-proto types almost everywhere. That's how I arrived at that idea/question of whether that's an intentional choice to only use them "as internals" in the ./lib/ packages.

Reply:
Agreed on having a defined-by-ourselves public VersionList for this purpose.

As for the larger thinking, I guess my reasons were something along the lines of
- The generated types are kind of ugly in shape (pointers and getters everywhere), so we currently don't want to use them primarily inside our code (that might change though) and we certainly don't want them to be part of our external interface, to the extent we have such a thing.
- Things we don't want other people to depend on should be in `internal`, lest we end up in the situation we are now, which I think is a mistake (that we could/should correct). Adding more stuff to `lib` that we don't want people to use would be to shoot ourselves again in the other foot.

So I guess it's a sneaky way to start slowly moving things to `internal` and see who screams.

Reply:
> Doing the opposite really requires us to freeze all interfaces in lib for compatibility, which is bullshit. So imho it's anyway all internal and has always been...

Fully agreed. I just consider some things less internal/more public-ish than others, and thinking about not unecessarily "internalizing"/breaking those (though we still can of course). And orthogonally trying to get a somewhat coherent model of what is internal by structure vs by intent. Thanks for the explanations for that!

Reply:
> Agreed on having a defined-by-ourselves public VersionList for this purpose. This one is just a debug method of some kind so we probably don't need to over-engineer it.

Clearly optional, I am likely missing why it's cumbersome to do this:  
You now added a special type for the debug version list, wrapping the proto version list. At the same time we also have a lot of utility functions for the proto version list, some of them duplicated between cmd (index) and lib/db. What about having a general `db.VersionList` type as the proto equivalent instead of the debug only type, such that all the version list utility functions become methods on that and can be shared between cmd and lib/db?

Reply:
Partly it's cumbersome because it has domino effects. Say we want a type for FileVersion so we can attach methods, nothing complicated, just `type FileVersion { *dbproto.FileVersion }` and off we go. However, FileVersions typically come from a VersionList, where we have lots of places where we iterate over `someVersionList.Versions` or append to it -- these now need to do wrapping, or we need to create accessor methods and conversion for the slice on VersionList. Alternatively we go the whole route of duplicating the type entirely and having marshalling functions on it...

Alternatively-alternatively we could generate the protobuf types right into the db package and attach methods directly like we used to do, but then we'd have to separate it into two separate protobuf packages because other types like Counts and ObserverFolder leak to the outside and their shape would change annoyingly everywhere... which also felt cumbersome. So here we are 😓